File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
int dp[2501][2501];
    int fun(vector<int>& nums,int i,int prev)
    {
        if(i>=nums.size())
        {
            return 0;
        }
        if(dp[i][prev]!=-1)
        {
            return dp[i][prev];
        }
        int skip=0,take=0;
        skip=fun(nums,i+1,prev);
        if(nums[prev]<nums[i])
        {
            take=1+fun(nums,i+1,i);
        }
        return dp[i][prev]=max(skip,take);
    }
    int lengthOfLIS(vector<int>& nums) {
        nums.insert(nums.begin(),-10001);
        memset(dp,-1,sizeof(dp));
        return fun(nums,0,0);
    }
};

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    private int helper(int[] nums,int i,int last,int[][] dp)
    {
        if(i==nums.length)
        {
            return 0;
        }
        if(dp[i][last]!=-1)
        {
            return dp[i][last];
        }
        int pick=0,skip=0;
        if(last==0)
        {
            pick=1+helper(nums,i+1,i+1,dp);
        }
        else if(nums[last-1]<nums[i])
        {
            pick=1+helper(nums,i+1,i+1,dp);
        }
        skip=helper(nums,i+1,last,dp);

         return dp[i][last]=Math.max(pick,skip);
    }
    public int lengthOfLIS(int[] nums) {
        int n=nums.length;
        int[][] dp = new int[n+1][n+1];
        for(int[] i : dp)
        {
            Arrays.fill(i,-1);
        }
        return helper(nums,0,0,dp);
    }
}